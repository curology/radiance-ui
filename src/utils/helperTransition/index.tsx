import React from 'react';
import { Transition } from 'react-transition-group';

import { ANIMATION } from '../../constants';

const START_STATE = {
  opacity: '0',
  maxHeight: '0',
};

const getStyleForTransitionState = (transitionState: string) => {
  switch (transitionState) {
    case 'entering':
      return START_STATE;
    case 'entered':
      return {
        opacity: '1',
        maxHeight: '48px',
        transitionDuration: ANIMATION.defaultTiming,
        transitionProperty: 'max-height, opacity',
        transitionTimingFunction: 'ease-in-out',
      };
    case 'exiting':
      return {
        opacity: '0',
        maxHeight: '0',
        transitionDuration: ANIMATION.defaultTiming,
        transitionProperty: 'max-height, opacity',
        transitionTimingFunction: 'ease-in-out',
      };
    case 'exited':
      return START_STATE;
    default:
      return {};
  }
};

interface HelperTransitionParamsType {
  children: JSX.Element;
  [key: string]: unknown;
}

const HelperTransition = ({
  children: child,
  ...props
}: HelperTransitionParamsType) => (
  // eslint-disable-next-line react/jsx-props-no-spreading
  <Transition timeout={350} {...props}>
    {(transitionState) =>
      React.cloneElement(child, {
        style: getStyleForTransitionState(transitionState),
      })
    }
  </Transition>
);

export default HelperTransition;
