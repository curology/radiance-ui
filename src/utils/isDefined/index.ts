function isDefined<T>(arg: T | null | undefined): arg is T {
  return arg !== undefined && arg !== null;
}

export { isDefined };
