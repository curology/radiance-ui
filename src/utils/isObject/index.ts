function isObject(
  val:
    | string
    | number
    | Record<string, unknown>
    | Array<unknown>
    | null
    | undefined
    | unknown,
): val is Record<string, unknown> {
  return val != null && typeof val === 'object' && !Array.isArray(val);
}

export default isObject;
