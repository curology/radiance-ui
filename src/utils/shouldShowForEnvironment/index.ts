const defaultAllowedEnvironments = ['test', 'development', 'dev'];

const environment = process.env.NODE_ENV;

const shouldShowForEnvironment = (
  allowedEnvironments = defaultAllowedEnvironments,
) =>
  typeof environment === 'string'
    ? allowedEnvironments.includes(environment)
    : false;

export default shouldShowForEnvironment;
