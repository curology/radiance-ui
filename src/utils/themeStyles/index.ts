/**
 *  Any conditional style based on theme.__type should be on this file
 */
import type { ThemeColors, ThemeType } from '../../constants';

export const primaryButtonFontColor = (theme: ThemeType) =>
  theme.__type === 'primary' ? theme.COLORS.white : theme.COLORS.primary;

export const primaryButtonBackgroundColor = (
  theme: ThemeType,
  buttonColor: ThemeColors,
) => {
  if (buttonColor === theme.COLORS.primary && theme.__type === 'secondary') {
    return theme.COLORS.secondary;
  }

  // If buttonColor is not COLORS.primary then it is custom, return as is
  return buttonColor;
};

export const primaryButtonLoadingBackgroundColor = (theme: ThemeType) =>
  theme.__type === 'primary' ? theme.COLORS.white : theme.COLORS.primary;

/**
 * We use theme.FONTS.baseFont for all primary styles, but use a
 * different secondary font for Display, Heading, and Title styles
 */
export const setSecondaryHeadingFont = (theme: ThemeType) =>
  theme.__type === 'secondary' ? `font-family: ${theme.FONTS.headerFont};` : '';

export const setButtonStyleFontWeight = (theme: ThemeType) =>
  theme.__type === 'primary'
    ? `font-weight: ${theme.TYPOGRAPHY.fontWeight.bold};`
    : '';

export const setThemeLineHeight = (
  theme: ThemeType,
  primaryLineHeight: string | number,
) => (theme.__type === 'secondary' ? '1.4' : primaryLineHeight);

export const setThemeFontWeight = (theme: ThemeType) =>
  theme.__type === 'secondary'
    ? theme.TYPOGRAPHY.fontWeight.normal
    : theme.TYPOGRAPHY.fontWeight.bold;

export const applyPrimaryThemeVerticalOffset = (
  theme: ThemeType,
  offset = '1',
) => (theme.__type === 'primary' ? `transform: translateY(${offset}px);` : '');
