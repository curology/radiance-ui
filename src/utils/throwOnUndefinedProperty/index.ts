export default function throwOnUndefinedProperty<
  T extends Record<string, unknown>,
>(obj: T): T {
  const handler = {
    get(target: T, property: string) {
      if (property === '__isProxy') {
        return true;
      }

      if (property in target) {
        return target[property];
      }

      // Allow React-Fast-Refresh to check if objects are components
      if (property === '$$typeof') {
        return;
      }

      throw new Error(`Property '${property}' is not defined`);
    },
  };

  return new Proxy(obj, handler);
}
