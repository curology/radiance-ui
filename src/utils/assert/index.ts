import { isDefined } from '../isDefined';

function assert(value: boolean, message?: string): asserts value;
function assert<T>(
  value: T | null | undefined,
  message?: string,
): asserts value is T;
/**
 * @see https://gist.github.com/jamiebuilds/2a40f565cba317a80bed1eec127340f6
 */
function assert(value: unknown, message?: string) {
  if (value === false || !isDefined(value)) {
    throw new Error(message ?? 'Assertion failed');
  }
}

export { assert };
