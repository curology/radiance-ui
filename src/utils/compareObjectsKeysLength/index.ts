import isObject from '../isObject';

// eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
interface RecursiveObjectType {
  [key: string]: string | number | this | null;
}

const stringifyObjects = (
  obj1: RecursiveObjectType,
  obj2: RecursiveObjectType,
) => `\n\nobj1: ${JSON.stringify(obj1)}\n\nobj2:${JSON.stringify(obj2)}`;

const compareObjectsKeysLength = (
  obj1: RecursiveObjectType,
  obj2: RecursiveObjectType,
): boolean => {
  if (Object.keys(obj1).length !== Object.keys(obj2).length) {
    console.error(`Object keys do not match:${stringifyObjects(obj1, obj2)}`);
    return false;
  }

  return Object.keys(obj1).every((key) => {
    const obj1Val = obj1[key];
    const obj2Val = obj2[key];

    // Test deep key equality recursively
    if (isObject(obj1Val) && isObject(obj2Val)) {
      return compareObjectsKeysLength(obj1Val, obj2Val);
    }

    // Check if property value is null or undefined, neither of which are valid
    // theme values, which means there was not a complementary key found
    if (obj2Val == null) {
      console.error(
        `Key "${key}" missing in obj2:${stringifyObjects(obj1, obj2)}`,
      );
      return false;
    }

    return true;
  });
};

export default compareObjectsKeysLength;
