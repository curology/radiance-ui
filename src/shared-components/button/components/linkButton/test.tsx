import React from 'react';
import { render } from 'src/tests/testingLibraryHelpers';
import { assert } from 'src/utils/assert';

import { LinkButton } from './index';

describe('<LinkButton/>', () => {
  describe('UI snapshots', () => {
    it('renders with props', () => {
      const { container } = render(
        <LinkButton onClick={() => undefined} href="#">
          Click me!
        </LinkButton>,
      );

      expect(container.firstElementChild).toMatchSnapshot();
    });
  });

  describe('href handling', () => {
    it('should link to a path', () => {
      const href = '/some/path';

      const { getByRole } = render(<LinkButton href={href}>text</LinkButton>);

      const anchorLink = getByRole('link') as HTMLAnchorElement;

      // Covers domain, e.g. http://localhost:8000/some/path
      expect(anchorLink.href.includes(href)).toBeTruthy();
    });
  });

  describe('onClick callback', () => {
    it('should be invoked onClick', async () => {
      const spy = jest.fn();

      const { container, user } = render(
        <LinkButton onClick={spy}>text</LinkButton>,
      );

      assert(container.firstElementChild);
      await user.click(container.firstElementChild);

      expect(spy).toHaveBeenCalled();
    });

    it('should not be invoked if disabled', async () => {
      const spy = jest.fn();

      const { container, user } = render(
        <LinkButton disabled href="#" onClick={spy}>
          text
        </LinkButton>,
      );

      assert(container.firstElementChild);
      await user.click(container.firstElementChild);

      expect(spy).not.toHaveBeenCalled();
    });
  });
});
