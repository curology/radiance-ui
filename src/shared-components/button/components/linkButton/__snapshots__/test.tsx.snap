// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LinkButton/> UI snapshots renders with props 1`] = `
.emotion-0 {
  color: #524D6E;
  font-size: 0.75rem;
  line-height: 1.67;
  font-weight: bold;
  letter-spacing: 1px;
  text-transform: uppercase;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  cursor: pointer;
  display: block;
  margin: 0;
  min-height: 52px;
  opacity: 1;
  padding: 0 1.5rem;
  position: relative;
  -webkit-transition: all 350ms ease-in-out;
  transition: all 350ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #332E54;
  border-color: #332E54;
  color: #FFFFFF;
  fill: #FFFFFF;
  min-width: 208px;
  max-width: 325px;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
}

.emotion-0:hover {
  -webkit-transition: all 350ms ease-in-out;
  transition: all 350ms ease-in-out;
}

.emotion-0:active,
.emotion-0:focus {
  outline: none;
  box-shadow: 0px 0px 0px 2px #FFFFFF,0px 0px 0px 4px #332E54;
}

.emotion-0:visited,
.emotion-0:hover {
  opacity: 0.8;
}

.emotion-0:focus,
.emotion-0:not([href]):not([tabindex]):hover,
.emotion-0:not([href]):not([tabindex]):focus {
  color: #FFFFFF;
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding: 1rem 0;
  -webkit-transition: -webkit-transform 350ms;
  transition: transform 350ms;
  width: 100%;
  -webkit-transform: translateX(0);
  -moz-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
}

.emotion-1>svg {
  opacity: 1;
  -webkit-transition: opacity 350ms;
  transition: opacity 350ms;
  margin-right: 1rem;
  margin-top: -5px;
}

.emotion-3 {
  line-height: 1.5;
  margin: 0;
  padding-top: 2px;
}

<a
  class="emotion-0"
  href="#"
>
  <div
    class="emotion-1 emotion-2"
  >
    <span
      class="emotion-3 emotion-4"
    >
      Click me!
    </span>
  </div>
</a>
`;
